### PR TITLE
[#353] Validation of external $ref property values should show error …

### PR DESCRIPTION
**Pull request by ghillairet**
…on unexpected object type

Validation of reference types has been moved to ReferenceValidator that previously only validate correctness of references pointer/URIs.
It now checks the type of the reference if this one is valid, this allow to reuse the code for loading the JSON document which is referenced by it.
This validation could be slow because it will parse the referenced documents into Models.

See #353

### CODE REVIEW

**Comment by ghillairet** [comment]
@tfesenko Can you please review. I updated this PR to include validation of types for external references. See #345 #353

File: com.reprezen.swagedit.openapi3.tests/src/com/reprezen/swagedit/openapi3/validation/ReferenceValidatorTest.xtend [hunk 1]
```diff
@@ -0,0 +1,393 @@
...
+			      in: query
+			      type: string
+		'''
+
+		document.set(content)
+		val baseURI = new URI(null, null, null)
+		val resolvedURI = new URI(null, null, "/components/parameters/Valid")
+		val errors = validator(#{resolvedURI -> document.asJson}).validate(baseURI, document)
+
+		assertEquals(0, errors.size())
```
Comment (tfesenko) [inline com.reprezen.swagedit.openapi3.tests/src/com/reprezen/swagedit/openapi3/validation/ReferenceValidatorTest.xtend:55]: Thanks!

File: com.reprezen.swagedit.core/src/com/reprezen/swagedit/core/json/references/JsonDocumentManager.java [hunk 2]
```diff
@@ -91,11 +91,6 @@ public JsonNode getDocument(URL url) {
     }
 
     public JsonNode getDocument(URI uri) {
-        final IFile file = getFile(uri);
-        if (file == null || !file.exists()) {
-            return null;
-        }
-
         try {
             return getDocument(uri.toURL());
```
Comment (tfesenko) [inline com.reprezen.swagedit.core/src/com/reprezen/swagedit/core/json/references/JsonDocumentManager.java:71]: We are passing an entire URL to `JsonDocumentManager.getDocument(URL)` which includes both the `file` and `ref` parts. `JsonDocumentManager.getDocument(URL)` has a table whose keys are `URL`s in a form we are passing the here. As a result, for two different refs from the **same** external model will load the model twice. For example, for the `beamup_path_estimates_price.yaml` from the Multi-file Beamup (v3) example the `documents` map will contain a different JsonNode model for each parameter:
```
{<...>/Multi-file%20Beamup%20(v3)/models/components/beamup_parameters.yaml#/components/parameters/start_longitude
    ={"openapi":"3.0.0", <...>}, 

<...>/Multi-file%20Beamup%20(v3)/models/components/beamup_parameters.yaml#/components/parameters/end_longitude
   ={"openapi":"3.0.0",<...>}
}
```
But it should be a single entry for `<...>/Multi-file%20Beamup%20(v3)/models/components/beamup_parameters.yaml`! 
We should fix `JsonDocumentManager.getDocument(URL)`.

File: com.reprezen.swagedit.core/src/com/reprezen/swagedit/core/json/references/JsonReference.java [hunk 3]
```diff
@@ -223,4 +216,19 @@ public static JsonPointer getPointer(ObjectNode node) {
         }
     }
 
+    /**
```
Comment (tfesenko) [inline com.reprezen.swagedit.core/src/com/reprezen/swagedit/core/json/references/JsonReference.java:220]: Javadoc for `JsonReference.isMissing()` should be corrected:
```java
    /**
     * Returns true if the reference points to an existing JSON document and the pointer to an existing node inside that
     * document.
     * 
     * @param baseURI
     * @return true if the reference can be resolved.
     */
    public boolean isMissing(JsonDocument document, URI baseURI) {
        if (isInvalid()) {
            return false;
        }

        JsonNode resolved = resolve(document, baseURI);
        return resolved == null || resolved.isMissingNode();
    }
```
It returns true if the reference can**not** be resolved and the pointer points to an **inexistent** element.

**Review by tfesenko — CHANGES_REQUESTED** [review]
I left several comments including a comment about the `documents` cache.
Also, we have several test failures on this branch:
```
Failed tests: 
  CodeTemplateTextTest.test_code_template_text:148 JSON Schema validation error: 

Tests in error: 
  ValidatorTest.testValidationShouldFail_ForInvalidPointers:314 » IllegalArgument
  ValidatorTest.testValidationShouldFail_refNotJson:368 » Scanner null; mapping ...
```

**Comment by ghillairet** [comment]
@tfesenko The PR has been updated.

**Comment by tfesenko** [comment]
@ghillairet , I am still getting two test failures on this branch (after merge master in to the task branch, the master is green):
```
Failed tests: 
  ValidatorTest.testValidationShouldFail_ForInvalidPointers:317 expected:<1> but was:<2>
  ValidatorTest.testValidationShouldFail_refNotJson:371 expected:<1> but was:<2>

```

**Comment by tfesenko** [comment]
Passed code review and QA.
All OK except for test failures.

**Comment by ghillairet** [comment]
Tests have been fixed.

File: com.reprezen.swagedit.openapi3.tests/src/com/reprezen/swagedit/openapi3/validation/ValidatorTest.xtend [hunk 4]
```diff
@@ -164,8 +164,8 @@ class ValidatorTest {
 			paths: {}
 			components: 
 			  schemas:
-			  
 			    MyType1:
+			      type: object
```
Comment (tfesenko) [inline com.reprezen.swagedit.openapi3.tests/src/com/reprezen/swagedit/openapi3/validation/ValidatorTest.xtend:168]: :smile:

**Comment by tfesenko** [comment]
Looks good. Thanks!!

**Pull request merged**